### ISSUE TITLE
`date_trunc` returning `null` only when a table is joined

### ISSUE THREAD

**Issue by bdalenoord** [timescale/timescaledb#2789]
**Relevant system information:**
 - OS: Docker on MacOS X, also on Ubuntu server
 - PostgreSQL version (output of `postgres --version`): 12.5
 - TimescaleDB version (output of `\dx` in `psql`): 2.0.0
 - Installation method: using Docker

**Describe the bug**
I've got an issue with `date_trunc` as soon as I join a table.

The following query proves to be no issue:
```sql
SELECT
    data.remote_timestamp AS deviationTimestamp,
    pg_typeof(data.remote_timestamp),
    date_trunc('hour', data.remote_timestamp),
    date_trunc('hour', data.remote_timestamp) IS NULL
FROM
    data
-- LEFT JOIN
--     averagesBucket ON
--     date_trunc('hour', data.remote_timestamp) = date_trunc('hour', averagesBucket.averageTimestamp)
WHERE
    data.remote_timestamp > NOW() - ('14 days')::interval
AND
    data.remote_timestamp < NOW()
ORDER BY
    data.remote_timestamp DESC
;
```
You see a commented-out `LEFT JOIN` here. If I execute this query, I get an output like this:
```
 deviationtimestamp  |          pg_typeof          |     date_trunc      | ?column?
---------------------+-----------------------------+---------------------+----------
 2021-01-04 14:58:00 | timestamp without time zone | 2021-01-04 14:00:00 | f
 2021-01-04 14:57:00 | timestamp without time zone | 2021-01-04 14:00:00 | f
 2021-01-04 14:56:00 | timestamp without time zone | 2021-01-04 14:00:00 | f
 2021-01-04 14:55:00 | timestamp without time zone | 2021-01-04 14:00:00 | f
 2021-01-04 14:54:00 | timestamp without time zone | 2021-01-04 14:00:00 | f
 2021-01-04 14:53:00 | timestamp without time zone | 2021-01-04 14:00:00 | f
 2021-01-04 14:52:00 | timestamp without time zone | 2021-01-04 14:00:00 | f
 2021-01-04 14:51:00 | timestamp without time zone | 2021-01-04 14:00:00 | f
 ...
 ```

As soon as I un-comment the LEFT JOIN, the situation changes. 
```sql
SELECT
    data.remote_timestamp AS deviationTimestamp,
    pg_typeof(data.remote_timestamp),
    date_trunc('hour', data.remote_timestamp),
    date_trunc('hour', data.remote_timestamp) IS NULL
FROM
    data
LEFT JOIN
    averagesBucket ON
    date_trunc('hour', data.remote_timestamp) = date_trunc('hour', averagesBucket.averageTimestamp)
WHERE
    data.remote_timestamp > NOW() - ('14 days')::interval
AND
    data.remote_timestamp < NOW()
ORDER BY
    data.remote_timestamp DESC
;
```
Using this query, I get the following output:
```
 deviationtimestamp  |          pg_typeof          | date_trunc | ?column?
---------------------+-----------------------------+------------+----------
 2021-01-04 14:58:00 | timestamp without time zone |            | t
 2021-01-04 14:57:00 | timestamp without time zone |            | t
 2021-01-04 14:56:00 | timestamp without time zone |            | t
 2021-01-04 14:55:00 | timestamp without time zone |            | t
 2021-01-04 14:54:00 | timestamp without time zone |            | t
 2021-01-04 14:53:00 | timestamp without time zone |            | t
 2021-01-04 14:52:00 | timestamp without time zone |            | t
 2021-01-04 14:51:00 | timestamp without time zone |            | t
 ...
```
You see the last two colums being blank, `date_trunc` returns `null`.

I'm able to recreate this issue in a new clean Docker-TimescaleDB-container (:latest-pg12), but not in an otherwise equivalent PostgreSQL-12.5-container, as PostgreSQL gives me the same output for both the query with and without the join (which in my opinion is the expected behavior).

**To Reproduce**
1. Spin  up a container: `docker run --rm --name timescale_test -p 1234:5432 -e POSTGRES_PASSWORD=password timescale/timescaledb:latest-pg12`
2. Initialize the database
```sql
BEGIN;
DROP TABLE IF EXISTS data;
DROP TABLE IF EXISTS averagesBucket;

CREATE TABLE data (remote_timestamp TIMESTAMP NOT NULL);
SELECT create_hypertable('data', 'remote_timestamp');

CREATE OR REPLACE FUNCTION generateDataForDate(
    date_to_generate date,
    minutes_to_generate int
) RETURNS VOID AS
$$
DECLARE
    minute_to_generate time;
    START_OF_DAY CONSTANT time := '00:00:00'::time;
BEGIN
    FOR minute IN 0..(minutes_to_generate - 1) LOOP
        minute_to_generate = START_OF_DAY + (minute || ' minute')::interval;
        INSERT INTO data(remote_timestamp) VALUES (date_to_generate + minute_to_generate);
    END LOOP;
END
$$ language 'plpgsql' STRICT;

DO
$BODY$
DECLARE
        number_of_days int;
        date_to_generate date;
        MINUTES_PER_DAY CONSTANT int := 1440;
BEGIN
    number_of_days = 50;
    FOR dayNum IN 1..number_of_days LOOP
        date_to_generate = CURRENT_TIMESTAMP::date - (number_of_days - (dayNum - 1) || ' day')::interval;
        RAISE NOTICE 'Generating data for date %', date_to_generate;
        EXECUTE generateDataForDate(date_to_generate, MINUTES_PER_DAY);
    END LOOP;

    RAISE NOTICE 'Generating data for date %', CURRENT_DATE;
    EXECUTE generateDataForDate(CURRENT_DATE, floor(extract(epoch FROM (localtime - '0:0'::time)/60))::int);
END
$BODY$
LANGUAGE plpgsql;

CREATE TABLE averagesBucket AS
SELECT
    time_bucket_gapfill('1 hour', remote_timestamp) AS averageTimestamp
FROM
    data
WHERE
    remote_timestamp > NOW() - ('14 days')::interval
AND
    remote_timestamp < NOW()
GROUP BY
    averageTimestamp
;
COMMIT;
```
3. Run the stated queries, observe the behavior

To test PostgreSQL as well, the initialization process is a bit different:
1. Spin up a container: `docker run --rm --name tspostgres_test -p 1235:5432 -e POSTGRES_PASSWORD=password postgres:12.5`
2. Initialize the database:
```sql
BEGIN;
DROP TABLE IF EXISTS data;
DROP TABLE IF EXISTS averagesBucket;

CREATE TABLE data (remote_timestamp TIMESTAMP NOT NULL);

CREATE OR REPLACE FUNCTION generateDataForDate(
    date_to_generate date,
    minutes_to_generate int
) RETURNS VOID AS
$$
DECLARE
    minute_to_generate time;
    START_OF_DAY CONSTANT time := '00:00:00'::time;
BEGIN
    FOR minute IN 0..(minutes_to_generate - 1) LOOP
        minute_to_generate = START_OF_DAY + (minute || ' minute')::interval;
        INSERT INTO data(remote_timestamp) VALUES (date_to_generate + minute_to_generate);
    END LOOP;
END
$$ language 'plpgsql' STRICT;

DO
$BODY$
DECLARE
        number_of_days int;
        date_to_generate date;
        MINUTES_PER_DAY CONSTANT int := 1440;
BEGIN
    number_of_days = 50;
    FOR dayNum IN 1..number_of_days LOOP
        date_to_generate = CURRENT_TIMESTAMP::date - (number_of_days - (dayNum - 1) || ' day')::interval;
        RAISE NOTICE 'Generating data for date %', date_to_generate;
        EXECUTE generateDataForDate(date_to_generate, MINUTES_PER_DAY);
    END LOOP;

    RAISE NOTICE 'Generating data for date %', CURRENT_DATE;
    EXECUTE generateDataForDate(CURRENT_DATE, floor(extract(epoch FROM (localtime - '0:0'::time)/60))::int);
END
$BODY$
LANGUAGE plpgsql;

CREATE TABLE averagesBucket AS
SELECT
    date_trunc('hour', remote_timestamp) AS averageTimestamp
FROM
    data
WHERE
    remote_timestamp > NOW() - ('14 days')::interval
AND
    remote_timestamp < NOW()
GROUP BY
    averageTimestamp
;
COMMIT;
```
3. Run the queries above, observe the correct behavior

**Expected behavior**
I expect `date_trunc` to return me a truncated date for the same given input

**Actual behavior**
`date_trunc` returns `null` given the same input when an arbitrary join is performed

**Comment by bdalenoord** [comment]
Note: the PostgreSQL-initialization is a bit different as it does not have the `create_hypertable` (of course) nor does it support the `time_bucket` so instead I use `date_trunc` there as well. I don't see why that would cause an issue, but if I'm wrong please let me know.

**Comment by svenklemm** [comment]
Closing since this should be fixed by #2860 which is included in 2.0.1. Feel free to reopen if it still occurs on 2.0.1.

**Comment by bdalenoord** [comment]
Confirmed this being fixed when using 2.0.1, nice!
